src/ObjectPool/src/DisposableObjectPool.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading;
+
+namespace Microsoft.Extensions.ObjectPool
+{
+    public class DisposableObjectPool<T> : DefaultObjectPool<T>, IDisposable where T : class
+    {
+        private volatile bool _isDisposed;
+
+        public DisposableObjectPool(IPooledObjectPolicy<T> policy)
+            : base(policy)
+        {
+        }
+
+        public DisposableObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
+            : base(policy, maximumRetained)
+        {
+        }
+
+        public override T Get()
+        {
+            if (_isDisposed)
+            {
+                ThrowObjectDisposedException();
+            }
+
+            return base.Get();
+
+            void ThrowObjectDisposedException()
+            {
+                throw new ObjectDisposedException(this.GetType().Name);
+            }
```

Comment:
Do this give the same code quality benefits as static throw helper? Curious because it captures `this`.

Reply:
Yeah, same benefit. The key here is that the compiler generated static method will just call `GetType`, so no `this` is needed, as the type can be also be obtained from the static method.

I'll remove the explicit `this`, as it is not needed and doesn't follow the coding conventions here.

<details>
  <summary>JIT produces</summary>

```asm
G_M36605_IG01:
       55                   push     rbp
       488BEC               mov      rbp, rsp
       90                   nop      

G_M36605_IG02:
       807F2900             cmp      byte  ptr [rdi+41], 0
       750E                 jne      SHORT G_M36605_IG05

G_M36605_IG03:
       48B8B80C359A1B7F0000 mov      rax, 0x7F1B9A350CB8

G_M36605_IG04:
       5D                   pop      rbp
       48FFE0               rex.jmp  rax

G_M36605_IG05:
       E87AFAFFFF           call     DisposableObjectPool`1:<Get>g__ThrowObjectDisposedException|3_0():this
       CC                   int3 
```
Even with tail-call for `base.Get` -- good JIT :smile:
</details>